controller/controller.go

```diff
@@ -0,0 +1,596 @@
+func (c *Controller) enqueueConfigMap(obj interface{}) {
+	var key string
+	var err error
+	if key, err = cache.MetaNamespaceKeyFunc(obj); err != nil {
+		utilruntime.HandleError(err)
+		return
+	}
+	c.workqueue.Add(key)
+}
+
+// getClusterAlertRules builds a list of rules from all the configmaps in the cluster
+func (c *Controller) getClusterAlertRules(mapList *corev1.ConfigMapList) *[]rulefmt.RuleNode {
+	var finalRules []rulefmt.RuleNode
+	for _, cm := range mapList.Items {
+		if c.isRuleConfigMap(&cm) {
+			cmRules := c.extractValues(&cm)
+			if len(cmRules) > 0 {
+				finalRules = append(finalRules, cmRules...)
+
+			}
+		}
+	}
+	return &finalRules
+}
+
+// getLogzioGrafanaAlerts builds a list of rules from all logz.io
+func (c *Controller) getLogzioGrafanaAlerts(folderUid string) (*[]grafana_alerts.GrafanaAlertRule, error) {
+	alertRules, ListLogzioRulesErr := c.logzioAlertClient.ListGrafanaAlertRules()
+	if ListLogzioRulesErr != nil {
+		return nil, ListLogzioRulesErr
+	}
+	// find all alerts inside prometheus alerts folder
+	var alertsInFolder []grafana_alerts.GrafanaAlertRule
+	for _, rule := range alertRules {
+		if rule.FolderUID == folderUid {
+			alertsInFolder = append(alertsInFolder, rule)
+		}
+	}
+	return &alertsInFolder, nil
+}
+
+// findOrCreatePrometheusAlertsFolder tries to find the prometheus alerts folder in logz.io, if it does not exist it creates it.
+func (c *Controller) findOrCreatePrometheusAlertsFolder() (string, error) {
+	folders, err := c.logzioFolderClient.ListGrafanaFolders()
+	if err != nil {
+		return "", err
+	}
+	envFolderTitle := fmt.Sprintf("%s-%s", c.envId, alertFolder)
+	// try to find the prometheus alerts folder
+	for _, folder := range folders {
+		if folder.Title == envFolderTitle {
+			return folder.Uid, nil
+		}
+	}
+	// if not found, create the prometheus alerts folder
+	grafanaFolder, err := c.logzioFolderClient.CreateGrafanaFolder(grafana_folders.CreateUpdateFolder{
+		Uid:   fmt.Sprintf("%s-%s", envFolderTitle, generateRandomString(randomStringLength)),
+		Title: envFolderTitle,
+	})
+	if err != nil {
+		return "", err
+	}
+	return grafanaFolder.Uid, nil
+}
+
+// extractValues extracts the rules from the configmap, and validates them
+func (c *Controller) extractValues(cm *corev1.ConfigMap) []rulefmt.RuleNode {
+
+	fallbackNameStub := createNameStub(cm)
+
+	var toalRules []rulefmt.RuleNode
+
+	for key, value := range cm.Data {
+		// try each encoding
+		// try to extract a rules
+		var rule rulefmt.RuleNode
+		var err error
+		err, rule = c.extractRules(value)
+		if err != nil {
+			errorMsg := fmt.Sprintf("Configmap: %s key: %s Error during extraction.", fallbackNameStub, key)
+			c.configmapEventRecorderFunc(cm, corev1.EventTypeWarning, ErrInvalidKey, errorMsg)
+		}
+
+		// Add unique name for the alert rule to prevent duplicate rules ([alert_name]-[configmap_name]-[configmap_namespace])
+		rule.Alert.Value = fmt.Sprintf("%s-%s-%s", cm.Name, cm.Namespace, key)
```

Comment:
What will happen if for some reason the same cm has 2 alerts with the same name?

Reply:
In our case we take the alert name from the configmap data keys, so its not possible to have duplicate keys in the same configmap, so the combitation of -> `<cm-name>-<namespace>-<cm-data-key>` will be unique